docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+reduces maintenance effort by reducing Gardener's custom code base, and enables subsequent reduction in ShootKapi
+compute cost.
+
+## Motivation
+Gardener uses a relatively complex approach for scaling shoot control plane `kube-apiserver` instances (ShootKapi
+hereafter). This need is a consequence of the following concerns:
+1. Gardener supplies K8s clusters of highly diverse compute scale and utilization - ranging from a couple of small
+   practically idle nodes, to hundreds of heavily utilized giants. Furthermore, the ratio of 'cluster compute capacity'
+   to 'ShootKapi load' substantially varies between different applications utilizing Gardener shoots. As a result,
+   compute resources required by ShootKapis of different shoots vary by two orders of magnitude.
+2. The compute cost of ShootKapi workloads constitutes a major part of Gardener's overall compute cost. Small relative
+   waste in sizing ShootKapis results in considerable absolute cost.
+3. Kube-apiserver does not scale well horizontally, in a sense that extra replicas come with substantial compute overhead.   
+
+In short, ShootKapi needs to be scaled:
+- Within a broad range
+- With reasonable accuracy, relative to actual resource consumption
+
+The implication is that ShootKapi requires both horizontal and vertical scaling. The existing implementation utilized 
+to that end is [HVPA].
+
+### Existing Solution
+[HVPA] strives to achieve a fusion of HPA and VPA recommendations:
+`HvpaRecommendedCapacity = k * HpaRecommendedCapacity + (1 - k) * VpaRecommendedCapacity`
+The underlying goal is to allow HPA and VPA to be driven by the same CPU and memory usage metrics, and overcome the
+HPA-VPA incompatibility known to exist in that mode.
+
+### Deficiencies of the Existing Solution
+HVPA strives to achieve a linear combination of HPA's and VPA's scaling policies:
+`HvpaRecommendedCapacity = k * HpaRecommendedCapacity + (1 - k) * VpaRecommendedCapacity`.
+However, there is a nonobvious obstacle before this intent: both HPA and VPA's policies are practically unavailable -
+HVPA does not know how HPA/VPA would scale the workload, as function of circumstances. The only information available
+is "how would HPA/VPA change the existing workload right now". So, in place of the original intent, HVPA resorts to:
+`dHvpaRecommendedCapacity = k * dHpaRecommendedCapacity + (1 - k) * dVpaRecommendedCapacity` (the 'd' symbols stand for
+"rate of change").
+
+Had HPA and VPA recommendations been always deterministically available, up to date, and with absolute accuracy,
+that would result in the originally intended behavior. Simply put, in a 3-D space where one coordinate is the real-world
+utilization data, and the other two are the horizontal and vertical scale suggested by the recommender, this is a line -
+a line which is a weighted average of the two lines which are individual HPA and VPA recommendations (Fig. 1).
```

Comment:
I've done many iterations on this section... I can explain reasonably simply the discrepancy in terms of superficial behaviour. But as soon as I get to motivating "why we can't just fix it, and are going for a new solution", there's a sharp step up in complexity.

Now that you made me look into it again, I wonder if the description would work better, if I dedicate the short first paragraph ("HVPA strives to achieve...") strictly to how if fails in terms of behaviour, instead of mixing in information about the underlying algorithm.

The discussion about the core algorithmic issue can then be placed in a separate section, so it can be easily skipped. Right now it is mixed with the concern which follows it ("HVPA is also organised around a concept of multiple scaling stages..."), which forces the user to read the algorithmic part, just so they can find where the next point starts.

I think I'll try that. I think it will only be a small improvement, but with no downsides, so it's still worth it.

Reply:
Marking this thread 'resolved' to keep the PR manageable, but don't hesitate to reopen if necessary. I broke down the section into several subsections, each with its own title. Hopefully makes it a bit easier to navigate and read selectively. Will be available with next push.